Polygon tool

ErikOstlund opened the pull request:
Package: `lib-classifier`

Closes #1299 


## Describe your changes:

This PR adds the `polygon` drawing tool to the available tools in the `front-end-monorepo`.
The `polygon` is created with a `points` attribute containing an array of x, y coordinate objects:

```
points: [
  {x: 345.18, y:160.65},
  {x: 350.3, y:165.89},
  {x: 355, y:170.42}
]
```

The README file will describe the features in detail.

<em>Helpful explanations that will make your reviewer happy:</em>
- What Zooniverse project should my reviewer use to review UX?

[Testing URL](https://localhost:8080/?env=production&project=16201&workflow=18675)

- What user actions should my reviewer step through to review this PR?

1. Click to add points
2. Undo (remove a point)
3. Close Path
4. Move a single point
5. Move the entire polygon shape
6. Delete a polygon

**See the polygon README file for details.**

- Which storybook stories should be reviewed?
- Are there plans for follow up PR’s to further fix this bug or develop this feature?

A basic unit test is part of this PR which ensures the polygon renders properly. Full unit tests and Storybook file will be in the next PR.

goplayoutside3 commented on packages/lib-classifier/src/plugins/drawingTools/components/Polygon/Polygon.js:
```diff
@@ -0,0 +1,174 @@
+    mark.shortenPath()
+  }
+
+  function handleClosePolygon(i) {
```
In Polygon component, `handleClosePolygon()` is only used for an initial point's `onPointerDown()` event. There’s a check in `handleClosePolygon()` for index === 0, but there’s no case where `handleClosePolygon()` is called with an index other than 0. I think `handleClosePolygon()` could simply call `mark.finish()` and doesn't need to take a parameter.

goplayoutside3 commented on packages/lib-classifier/src/plugins/drawingTools/components/Polygon/Polygon.js:
```diff
@@ -0,0 +1,174 @@
+      {active &&
+        points.map((point, i) => {
+          if (i === 0) {
+            {
```
```suggestion
```

These are redundant brackets around the in-code comments

goplayoutside3 commented on packages/lib-classifier/src/plugins/drawingTools/components/Polygon/Polygon.js:
```diff
@@ -0,0 +1,174 @@
+          if (i === 0) {
+            {
+              /* Initial Point */
+            }
```
```suggestion
```

goplayoutside3 commented on packages/lib-classifier/src/plugins/drawingTools/components/Polygon/Polygon.js:
```diff
@@ -0,0 +1,174 @@
+            )
+          }
+
+          {
```
```suggestion
```

goplayoutside3 commented on packages/lib-classifier/src/plugins/drawingTools/components/Polygon/Polygon.js:
```diff
@@ -0,0 +1,174 @@
+
+          {
+            /* All other points in the Polygon */
+          }
```
```suggestion
```

goplayoutside3 reviewed (commented):
## PR Review

✔️ I’m able to load the test project, add points, undo points, close a path, move a single point when the path is open or closed, move the polygon shape when open or closed, and delete the polygon. 
✔️ I cannot undo points once the polygon is closed.
✔️ The grid area and polygon scale as I change my browser size.
🚨 I can create a finished, unclosed polygon. Is that intentional? To do so, I create points of an unclosed polygon, move it twice, and then click an empty spot on the grid intending to add another point, but the polygon is deselected as if finished. If that happens, I’m able to re-select the polygon and continue where I left off. I was also able to click ‘Done’ with an unclosed polygon drawn.
🚨 A note about UndoButton that I can detail in a separate issue if preferred. The UndoButton svg element is styled with precise pixels designed to fit only one screen size, so the arrow appears uncentered to any other screen. To correct this, we could create an svg path that includes both the circle and arrow so the arrow is always centered. Or create the containing circle button out of a dom element, and then center the arrow svg inside it rather than using the `undoTransform` pixel manipulation.

ErikOstlund commented:
- `handleClosePolygon` refactored
- `Undo` button is now a single svg
- Fixed move issue: Users can now add more points to an in-progress polygon after the polygon has been moved
- Removed curly brackets around commented code - had to adjust local prettier settings... so weird!

ErikOstlund commented:
@goplayoutside3 and @shaunanoordin 
NOTE: A known bug still exists but will be addressed in a separate PR because the issue stems from parent components (InteractionLayer and DrawingToolMarks). This issues needs to be kept isolated from this initial Polygon Drawing Tool PR.
The bug: Using the Polygon tool, if a user creates two points then changes tools (lets say a rectangle tool), the Polygon will remain on the drawing canvas as a line. This is an invalid Polygon shape. A valid polygon must have at least 3 points. What should happen: when the rectangle tool is selected, the invalid polygon drawing is deleted.

goplayoutside3 reviewed (approved):
## PR Review

Looks good! The UndoButton styling is consistent, and I can move an in-progress polygon. I tested each of the actions in the PR description:
- Click to add points
- Undo (remove a point)
- Close Path
- Move a single point
- Move the entire polygon shape
- Delete a polygon

Tests are passing locally too 👍

shaunanoordin commented on packages/lib-classifier/src/helpers/roundCoordinates/README.md:
```diff
@@ -0,0 +1,42 @@
+# roundCoordinates
+
+A helper function to round a set of { x, y } coordinates to two decimal places.
```
Docccuuumeeenntaaattiiioooonnnn! 👍

eatyourgreens replied:
Not needed for this PR, but I think this could be used as the basis of a custom type for rounded numbers.

shaunanoordin commented on packages/lib-classifier/src/plugins/drawingTools/models/tools/Tool/Tool.js:
```diff
@@ -90,6 +90,10 @@ const Tool = types
       mark.initialDrag?.(event)
     },
 
+    // this outputs the mouse coords when not creating (ig: guideline for Polygon)
+    // By default, this does nothing. Must be implemented in model/tool file (polygonTool.js)
+    handlePointerPosition(event, mark) {},
+
```
[documentation] I strongly recommend that you also update https://github.com/zooniverse/front-end-monorepo/blob/master/packages/lib-classifier/src/plugins/drawingTools/README.md with the new `handlePointerPosition` info.

But 👍 for having inline documentation too!

ErikOstlund replied:
Added ✅

shaunanoordin commented on packages/lib-classifier/src/plugins/drawingTools/models/marks/Polygon/Polygon.js:
```diff
@@ -0,0 +1,120 @@
+    }
+  }))
+  .actions((self) => ({
+    initialDrag() {},
```
Good to see this has been manually added, as per the [drawing tools documentation/intended design patterns](https://github.com/zooniverse/front-end-monorepo/blob/master/packages/lib-classifier/src/plugins/drawingTools/README.md) we were talking about. Nice! 👍

shaunanoordin commented on packages/lib-classifier/src/plugins/drawingTools/models/marks/Mark/Mark.js:
```diff
@@ -102,6 +102,10 @@ const BaseMark = types
       self.finished = true
     },
 
+    continueDrawing() {
+      self.finished = false
+    },
+
```
❓ [major] Should this be in the base Mark model, or the Polygon Mark model? I think it should be shifted down to the Polygon Mark, right?

ErikOstlund replied:
I'm thinking it should stay in `Mark` model because `finished` is created and initialized in the `Mark` model. Also, keeping it in the `Mark` model will future-proof this for any other tools that may need it.
Thoughts?

shaunanoordin replied:
Future-proofing is a good point, actually.

I still feel that `continueDrawing` is something that should go in the `Polygon` mark (and any future Polygon-like mark) though, since `mark.continueDrawing()` is _only_ called by the `Polygon` component (and any future Polygon-like component) and not by the InteractionLayer nor anything _shared by all types of Marks.

On a scale of 0 (ambivalent) to 10 (pistols at dawn, good sir, I will duel to defend my opinion), I feel like I'm a... 5 (I will tweet angrily but won't actually do anything #continueDrawingInPolygon) on this.

If you feel more strongly than I do, then keep is as is - all I ask is that you update the Mark's section in the plugins/drawingTools/README accordingly. 👌

shaunanoordin reviewed (changes requested):
## PR Review

This is an additional PR to supplement Delilah's review 👍 

Package: `lib-classifier`

This PR adds the Polygon Tool to FEM's Drawing Task:

- The intent of a Polygon Mark is to allow users to create a CLOSED polygon with at least 3 points/nodes.
- A Polygon Mark is created by clicking on the Subject image to place the first point/node, then the second, third, etc. 
  - The creation process is completed by the user clicking on the 
- Once a Polygon is finished, its nodes can be dragged to move the node.
- Once a Polygon is finished, its edges can be dragged to move the entire polygon.
- Standard behaviour: a Polygon mark can be deleted, of course.
- Polygon coordinates are rounded off to 2 decimal places, to prevent excessively large Classification sizes.

❗ This PR also makes some small changes to the **code Drawing system itself.** Please take note:

- Drawing Tools now have the `handlePointerPosition()` function.
  - On that note, InteractionLayer also now keeps track of the pointer position even when it's not in "creating" mode.
- DrawingToolMarks no longer calls `onFinish()` when a Mark has been moved.
  - ❌ [major] This change needs to be reverted, as we need that `onFinish()` to delete marks that are moved outside the bounds of the Subject image. Sorry for not realising this earlier.
- The Undo button styling has been tweaked, I think?

The Polygon tool itself looks excellent, and my comments and CRs tend to revolve the baseline drawing systems that have been touched. As always, while A-OK with anything happening in the Polygon/Freehand Line/Circle/etc tool, I'm particularly cautious whenever we make changes to the base Mark or Tool classes.

### Testing

No additional functionality/user testing. This supplementary PR review is focused on the code read.

### Status

Changes Requested:

- [x] The Drawing Tools [README](https://github.com/zooniverse/front-end-monorepo/blob/master/packages/lib-classifier/src/plugins/drawingTools/README.md) should be updated with info on `handlePointerPosition`
- [x] ❗ `DrawingToolMarks.js` and `DrawingToolMarks.spec.js` must be reverted
  - We can open an issue and solve the "can't move Polygon while creating" bug separately from this PR.  
- [ ] ❓ `Mark.continueDrawing()` should be moved down to `PolygonMark.continueDrawing()`, I think?
  - Please try this and let me know if that doesn't work, because if it doesn't, we need to reevaluate the need to change the base Mark model.

I think this PR will be good to go once these major CRs are addressed. Again, the Polygon tool looks good, it's the base drawing systems we need to be careful about. Also, bonus 👍 points for all the documentation! Always happy to see that.

ErikOstlund commented on packages/lib-classifier/src/components/Classifier/components/SubjectViewer/components/InteractionLayer/components/DrawingToolMarks/DrawingToolMarks.spec.js:
```diff
@@ -93,39 +104,37 @@ describe('Components > DrawingToolMarks', function () {
         expect(tool.marks.size).to.equal(1)
-        expect(onFinish).to.have.been.calledOnce()
-        expect(onDelete).to.not.have.been.called()
+        expect(onFinish).to.not.have.been.called()
```
Reverted

ErikOstlund commented on packages/lib-classifier/src/components/Classifier/components/SubjectViewer/components/InteractionLayer/components/DrawingToolMarks/DrawingToolMarks.js:
```diff
@@ -59,8 +59,6 @@ function DrawingToolMarks({
     function endMoveMark(event) {
       if (event?.currentTarget && !isInBounds(event.currentTarget)) {
         deleteMark()
-      } else {
-        onFinish(event)
```
Reverted

ErikOstlund commented on packages/lib-classifier/src/components/Classifier/components/SubjectViewer/components/InteractionLayer/components/DrawingToolMarks/DrawingToolMarks.js:
```diff
@@ -44,7 +44,7 @@ function DrawingToolMarks({
     }
 
-    function moveMark(event, difference) {
+    function moveMark(_event, difference) {
```
This is an update to remove a lint error.
`moveMark()` accepts two parameters: `event` and `difference`. Only `difference` is used in the function. This leaves `event` unused and throws `'event' is declared but its value is never read` lint error.
Placing an underscore in front of an unused parameter says, 'just ignore me and don't throw an error - I know I'm not being used'.

ErikOstlund commented on packages/lib-classifier/src/components/Classifier/components/SubjectViewer/components/InteractionLayer/InteractionLayer.js:
```diff
@@ -127,16 +127,20 @@ function InteractionLayer({
         activeTool.handlePointerMove(convertEvent(event), activeMark)
+    } else {
+      // this outputs the mouse coords when not creating (ig: guideline for Polygon)
+      activeTool.handlePointerPosition &&
+        activeTool.handlePointerPosition(convertEvent(event), activeMark)
```
Updated!

shaunanoordin reviewed (approved):
## PR Review (Update)

Changes LGTM! 👍 

I only have that [one comment](https://github.com/zooniverse/front-end-monorepo/pull/2883#discussion_r840784491) about you keeping `continueDrawing` in `Mark`. If you strongly believe that [this is the way,](https://www.youtube.com/watch?v=uelA7KRLINA), then all I ask is that you update the README for drawingTools accordingly.

Let's goooo! 👍

shaunanoordin commented:
Just adding a note here about what me & Erik were talking about on Slack:

You'll notice that in [c87eb60](https://github.com/zooniverse/front-end-monorepo/pull/2883/commits/c87eb600b9eeb3ea24ef4a718177104f596d934a), Erik **removed** _continueDrawing()_ altogether. Turns out, this bit of code was entirely unnecessary!

Prior to that commit, this was what was happening:
- `continueDrawing()`'s main job was to set the _finished mark_ to _UNfinished,_ which (in the case of the Polygon) causes the mark to be "drawable" again and _causes the Undo button to become visible._
- `continueDrawing()` can only be triggered _when the Undo button is already visible._
- But the Undo button _isn't visible on a finished mark._
- Therefore, you can't logically trigger `continueDrawing` on a finished mark, meaning you can't ever turn it into an _UNfinished_ mark.

As a result, this function was found safe to remove. 👌
